substrate/frame/assets/src/types.rs

```diff
@@ -232,6 +232,32 @@ impl<AssetId, AccountId, Balance> FrozenBalance<AssetId, AccountId, Balance> for
 	fn died(_: AssetId, _: &AccountId) {}
 }
 
+/// Trait for specifying a balance that is distinct from the free balance.
+/// This balance is then summed up with the balance of the account, and the
+/// `minimum_balance` (and frozen balance, if any) of the asset to calculate
+/// the reducible balance.
+pub trait HeldBalance<AssetId, AccountId, Balance> {
+	/// Return the held balance.
+	///
+	/// If `Some`, it means some balance is suspended, and it can be infallibly
+	/// slashed.
+	///
+	/// If `None` is returned, then nothing special is enforced.
+	fn held_balance(asset: AssetId, who: &AccountId) -> Option<Balance>;
```

Comment:
Should be `balance_on_hold`.

The vocabulary is "*Hold* this amount on the account", "The account has some balance *on hold*", "*Release* this balance which is *on hold*".

Reply:
Agreed.

There were two reasons behind picking this naming:

1. To preserve analogy with the `frozen_balance` method from `FrozenBalance` (that is also why the trait is called `HeldBalance` and not `BalanceOnHold`).
2. To ensure there is no naming clash with the `hold::Inspect::balance_on_hold` method. Otherwise, there might be (and would be easily found in some tests I wrote in case I rename this method) scenarios where a developer would be forced to explicitly cast the trait before calling the method to properly distinguish between them. An example of this is the following:
    ```rs
    <Pallet<T, I> as hold::Inspect<_>>::balance_on_hold(ASSET_ID, &who);
    <Pallet<T, I> as HeldBalance<_, _, _>>::balance_on_hold(ASSET_ID, &who);
    ```

Reply:
1. Held balance would mean it was on hold.
2. This is same as `frozen_balance` returning a `Balance` in fungibles but `Option<Balance>` in the `FrozenBalance` trait. The misleading is not related to a difference in concept but related to a difference in architecture: fungible makes no difference between no balance on hold and zero balance on hold. `BalanceOnHold` somehow makes a difference (though arguably it seems unnecessary at a first glance).

IMO BalanceOnHold is more sensible.

Reply:
Change applied on commit [818e590](https://github.com/paritytech/polkadot-sdk/pull/4530/commits/818e59048049ce51c6c3dca860653e533582769a)